### DISCUSSION TITLE
Gather ideas of how to shape the Dockerfile for alltalk beta

### DISCUSSION THREAD

**Discussion by Paladinium** [erew123/alltalk_tts#378] [Dev Discussions]
I would like to start a discussion of how to build a Dockerfile for alltalk beta, but also what the requirements are. This should support the process of crafting actual Dockerfiles.

> ℹ️ Update: A first draft is now available - see [this comment](https://github.com/erew123/alltalk_tts/discussions/378#discussioncomment-12348600)

### Goal of having Docker
The advantages of Docker are quite obvious: There is no need for local installations and dealing with dependencies and incompatibilities. Instead, a Docker image should provide everything needed to run alltalk. Also, it makes it easier to ship alltalk to production.

### Requirements
Just to give an idea of what kind of requirements are possible:

- When starting the image using `docker run`, alltalk should be up and running without the need for downloading dependencies. Startup time should therefore be in the order of seconds.
- All dependency versions should be pinned for the sake of reproducible builds (e.g. avoid `conda install -y pytorch::faiss-cpu` and instead use `conda install -y pytorch::faiss-cpu=1.9.0`)
   - On a running docker environment, complete requirements files for [pip and conda can separately be extracted](https://stackoverflow.com/questions/50777849/from-conda-create-requirements-txt-for-pip3) and added as additional alltalk requirements files.
   - This should allow for `--no-deps` to be used
- The docker images should be published to docker hub such that own Dockerfiles can be written using e.g. `FROM alltalk/alltalk:2.0`
- In turn, when building the image using `docker buildx build`, there should be the possibility of "configuring" the image, e.g. by passing [build args](https://docs.docker.com/build/building/variables/), e.g. for picking the model to use or providing some other means of configuration, e.g. by mounting config files, without having to modify the actual source code of alltalk.
   - It might be required only for internal purposes, e.g. when publishing docker images per model, e.g. `alltalk-xtts:2.0` or `alltalk-piper:2.0`
- The ability to use GPUs should be provided out-of-the box
- Extended features should be enabled/available by default such as
   - [DeepSpeed](https://github.com/erew123/alltalk_tts/wiki/Install-%E2%80%90-LINUX-%E2%80%90-Requirements-and-DeepSpeed)
   - [RVC](https://github.com/erew123/alltalk_tts/wiki/RVC-(Retrieval%E2%80%90based-Voice-Conversion))
   - [Multi-Engine-Manager](https://github.com/erew123/alltalk_tts/wiki/Multi-Engine-Manager)
- Adhere to best practices, e.g. 
   - Security: not running the application itself as root
   - Layering: Make sure Docker layers are cacheable (which they are not when running scripts)
- Forward logs to console
- ...

### Misc
Feel free to provide examples of Dockerfiles or state your further/other requirements.

**Comment by Paladinium** [comment]
An example of a Dockerfile which packages alltalk and picking xtts as model, yet not fulfilling many of the requirements, is:

```
FROM continuumio/miniconda3:24.7.1-0

SHELL ["/bin/bash", "-l", "-c"]
ENV SHELL=/bin/bash
ENV HOST=0.0.0.0
ENV DEBIAN_FRONTEND=noninteractive
ENV CUDA_DOCKER_ARCH=all

RUN <<EOR
apt-get update
# apt-get upgrade -y
apt-get install --no-install-recommends -y libaio-dev espeak-ng ffmpeg gcc g++ portaudio19-dev \
  git curl wget expect \
  freeglut3-dev vim

apt-get clean && rm -rf /var/lib/apt/lists/*
EOR

WORKDIR /alltalk

#RUN conda init bash
ARG INSTALL_ENV_DIR=/alltalk/alltalk_environment/env
ENV CONDA_AUTO_UPDATE_CONDA="false"
RUN <<EOR
    conda create -y -n "alltalk" -c conda-forge python=3.11.9
    conda activate alltalk
    conda install -y pytorch=2.2.1 torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia
    conda install -y nvidia/label/cuda-12.1.0::cuda-toolkit=12.1
    conda install -y pytorch::faiss-cpu
    conda install -y conda-forge::ffmpeg
    conda clean -a && pip cache purge
EOR

# Install python dependencies (cannot use --no-deps because requirements are not complete)
COPY system/config system/config
COPY system/requirements/requirements_standalone.txt system/requirements/requirements_standalone.txt
COPY system/requirements/requirements_parler.txt system/requirements/requirements_parler.txt
ENV PIP_CACHE_DIR=/alltalk/pip_cache
RUN <<EOR
    conda activate alltalk

    mkdir /alltalk/pip_cache
    pip install --no-cache-dir --cache-dir=/alltalk/pip_cache -r system/requirements/requirements_standalone.txt
    pip install --no-cache-dir --cache-dir=/alltalk/pip_cache --upgrade gradio==4.32.2

    # Deepspeed:
    curl -LO https://github.com/erew123/alltalk_tts/releases/download/DeepSpeed-14.0/deepspeed-0.14.2+cu121torch2.2-cp311-cp311-manylinux_2_24_x86_64.whl
    pip install --no-cache-dir --cache-dir=/alltalk/pip_cache deepspeed-0.14.2+cu121torch2.2-cp311-cp311-manylinux_2_24_x86_64.whl
    rm -f deepspeed-0.14.2+cu121torch2.2-cp311-cp311-manylinux_2_24_x86_64.whl

    # Parler:
    pip install --no-cache-dir --no-deps --cache-dir=/alltalk/pip_cache -r system/requirements/requirements_parler.txt

    conda clean --all --force-pkgs-dirs -y && pip cache purge
EOR

# Writing scripts to start alltalk:
RUN <<EOR
    cat << EOF > start_alltalk.sh
#!/usr/bin/env bash
source ~/.bashrc
conda activate alltalk
python script.py
EOF
    cat << EOF > start_finetune.sh
#!/usr/bin/env bash
source ~/.bashrc
export TRAINER_TELEMETRY=0
conda activate alltalk
python finetune.py
EOF
    cat << EOF > start_diagnostics.sh
#!/usr/bin/env bash
source ~/.bashrc
conda activate alltalk
python diagnostics.py
EOF
    chmod +x start_alltalk.sh
    chmod +x start_environment.sh
    chmod +x start_finetune.sh
    chmod +x start_diagnostics.sh
EOR

COPY . .

# Create scripts to execute firstrun.py:
# Options are: 1 = piper, 2 = vits, 3 = xtts
# TODO: Pass this option as ARG to the Dockerfile
RUN echo $'#!/usr/bin/env bash \n\
source ~/.bashrc \n\
conda activate alltalk \n\
python ./system/config/firstrun.py' > ./start_firstrun.sh
RUN echo $'#!/usr/bin/env expect \n\
expect_before timeout { exit 1 } \n\
set timeout 600 \n\
spawn ./start_firstrun.sh \n\
expect "Enter your choice" \n\
send -- "3\r" \n\
expect eof' > ./start_firstrun_noninteractive.sh

# Run start_firstrun_noninteractive.sh:
RUN chmod +x start_firstrun.sh start_firstrun_noninteractive.sh
RUN ./start_firstrun_noninteractive.sh

RUN mkdir -p /alltalk/outputs
RUN mkdir -p /root/.triton/autotune

# Enabling deepspeed for all models (TODO: isn't there a config option?):
RUN find . -name model_settings.json -exec sed -i -e 's/"deepspeed_enabled": false/"deepspeed_enabled": true/g' {} \;

## Start alltalk exposing the API port:
RUN chmod +x start_alltalk.sh
EXPOSE 7851
ENTRYPOINT ["sh", "-c", "./start_alltalk.sh"]
```

This can be built with `docker buildx build -t alltalk:latest .` and can be run with e.g. `docker run -p 7851:7851 --gpus=all -it alltalk:latest`

**Comment by erew123** [comment]
Hi @Paladinium What you have put above sounds great :) and I would love to have your help to do this. However, I have a small bit of code to complete withing the script.py code so that it identifies the Gradio interface is running within Docker. The API address will work fine, but the Gradio interface has to be pointed to the external IP address of wherever the API address is running from. If you are running docker on a local machine, it should (I think) work, but if you are running it somewhere on the internet in the cloud, it wont work.

Ive already done Google Colab, so that automatically picks up the tunnel addresses and pushes the IP/URL into the code as necessary when AllTalk starts up and I am kind of doing the same with when it detects a docker style environment. I have about 95% of the code written (somewhere on my home machine) but then this happened [link here](https://github.com/erew123/alltalk_tts/issues/377#text-generation-webui-users) a couple of months ago and I have been traveling away from home for a couple of months now. 

I still dont have a timeline when I will be back and able to finish it, it could be a week, it could be a month and I suppose there could be other possible timelines too (will make sense when you read the link).

I think I will be updating the Pytorch version in future, which may knock a few other requirements versions in the requirements files, but I don't think that would cause you any blips. However, please see my notes about the transformers versions on the link I gave above.  Its out of my hands getting a later version of transformers for Coqui TTS streaming to work, its handled by idiap, but they do appear to be discussing it/working on it https://github.com/idiap/coqui-ai-TTS. Also, Microsoft are finally starting to release pre-built wheels for Pytorch, so we may be able to do a generic link to just `pip install deepspeed` in future, but chatting with Logan at MS, they are currently having a few issues and of course, it will also depend on the Pytorch version used, so thats a drive to update it one day soon.

So, you **should** be able to get a docker build running on the API and call it with Command prompt CURL/SillyTavern Extension/TGWUI Remote Extension/, but the Gradio interface wont work *yet**. I am not great on Docker at all, but my brief glance says what you have done above and suggest looks good to me! :)

Hopefully I can get back to you soon, or maybe remote into my home machine and try check the code and post it up, though I have to be careful to make sure it doesnt knock something else out of whack and Im not exactly on the best laptop in the world to be able to remote control my home machine for coding. :/

## If anyone else has any other thoughts/ideas and would like to get involved in this chat, it would be great! 

And thanks again @Paladinium

**Reply by Paladinium** [reply]
Thanks for your feedback. Yes, the API calls are already working with the Dockerfile posted. I did not try any UIs because I don't need them.

In general, it is not possible for a single docker container to know the external IP address. Docker can be used in complex environments where there are many other components involved in networking (e.g. Kubernetes, load balancers, proxies, ...). I have never seen the need in a UI to have such information.

That said, Gradio itself has [instructions of how to use Gradio with Docker](https://www.gradio.app/guides/deploying-gradio-with-docker): similar to using `ENV HOST=0.0.0.0`, one can use `ENV GRADIO_SERVER_NAME="0.0.0.0"`.

**Reply by erew123** [reply]
Sounds good! :)

What I am doing in the gradio interface is when it detects Docker, there will be an additional box appears in the Gradio TTS generate page, so that you can point it to the API address/URL. Im sure youve figured that the TTS Engines/API and the Gradio interface are 2x different Python processes (and have to remain that way). 

Gradio should work for updating any config files via the interface, but not for API calls. So thats pretty much the work I have to complete. Will let you know back here when I have that code available and up.

**Reply by erew123** [reply]
@Paladinium Ive managed to finish the updates for Gradio allowing you to manually set the docker IP/URL in the interface. What I may also do at some point is allow a choice on startup in docker for Gradio's public shared address.

However, for now, if I have gotten things correct (which I havnt tested fully in docker if Im honest) you should be able to update the Gradio interface to point to the Docker IP/URL for the API, when you know what it is:

https://github.com/erew123/alltalk_tts/commit/0c9034c1f32a65573c3c4df4f9c907cd32b19be4

![image](https://github.com/user-attachments/assets/8566b616-6f6b-48b7-a047-e840142ffb07)

Sorry Ive not gotten to the PR yet, things are on/off very busy for me outside of all this and when I come back to do/test bits, there are support ticket requests etc to deal with. Will be on to it shortly though.

Thanks

**Reply by Paladinium** [reply]
Thanks. I will prioritize the changes regarding RVC first and then come back to the Dockerfile.

**Reply by vlrevolution** [reply]
> @Paladinium Ive managed to finish the updates for Gradio allowing you to manually set the docker IP/URL in the interface. What I may also do at some point is allow a choice on startup in docker for Gradio's public shared address.
> 
> However, for now, if I have gotten things correct (which I havnt tested fully in docker if Im honest) you should be able to update the Gradio interface to point to the Docker IP/URL for the API, when you know what it is:
> 
> [0c9034c](https://github.com/erew123/alltalk_tts/commit/0c9034c1f32a65573c3c4df4f9c907cd32b19be4)
> 
> ![image](https://private-user-images.githubusercontent.com/35898566/380460298-8566b616-6f6b-48b7-a047-e840142ffb07.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE5Njk1NjQsIm5iZiI6MTczMTk2OTI2NCwicGF0aCI6Ii8zNTg5ODU2Ni8zODA0NjAyOTgtODU2NmI2MTYtNmY2Yi00OGI3LWEwNDctZTg0MDE0MmZmYjA3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE4VDIyMzQyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTllZWVlYmZjNTkyZTI3MTNkNzJlOGJlYzhkNTQ2ZjNjZmEyZTBmMTJiZjBmMzQ5ZTk3ZmYxODRmYTM4ZjZkMGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.U4LpyzTMsinrUW3P-R2hsh6xNruYEF5RKQ-niYi9qWY)
> 
> Sorry Ive not gotten to the PR yet, things are on/off very busy for me outside of all this and when I come back to do/test bits, there are support ticket requests etc to deal with. Will be on to it shortly though.
> 
> Thanks

This setting for docker ip/host usign ngrok testing it in a vast.ai machine is not currently fixing the access issue. Maybe there's something wrong still?

**Reply by Paladinium** [reply]
@vlrevolution : Did you use the example I posted as 1st comment in this discussion using the code base from the `alltalkbeta` branch?

Just to be clear: This was an example and not yet fully completed. I was busy contributing at other places in this project, so no progress so far.

**Reply by vlrevolution** [reply]
> @vlrevolution : Did you use the example I posted as 1st comment in this discussion using the code base from the `alltalkbeta` branch?
> 
> Just to be clear: This was an example and not yet fully completed. I was busy contributing at other places in this project, so no progress so far.

No, I was talking about being able to change the docker ip in the panel, the image that @erew123 posted in his reply. I think right now if it detects docker it breaks the url it sends the api requests to in gradio, even if you change the ip in the panel (has no effect). I had to comment out the docker detection code to get it to use even the correct localhost ip (using ssh tunneling).

**Reply by erew123** [reply]
@vlrevolution The new update should (I say should) work correctly with docker, as far as the Gradio interface goes. I have not tested in docker, only faked the flag in the code to say its in docker and test that it works.

Obviously though @Paladinium will probably be testing a docker build at some point in the future.

**Reply by vlrevolution** [reply]
> @vlrevolution The new update should (I say should) work correctly with docker, as far as the Gradio interface goes. I have not tested in docker, only faked the flag in the code to say its in docker and test that it works.
> 
> Obviously though @Paladinium will probably be testing a docker build at some point in the future.

Thank you, I will test in near future and report back

**Comment by erew123** [comment]
I came across this for optimising docker image sizes. Its supposedly good, but I don't know enough about docker to say how easy this would be to use or what impact it may have. I'm leaving it here for reference as I'm happy to give it a shot when a docker image gets built https://github.com/slimtoolkit/slim

Claims of up to 30x reduction in size

**Comment by vlrevolution** [comment]
Looking at issues tho shouldnt get hopes up. AI docker images are already
very hard to get working in the first place

**Comment by RenNagasaki** [comment]
@vlrevolution it sounds like you managed to get a docker image working, if yes, would you mind sharing that one with us? I'd love to test that. 😄

**Reply by vlrevolution** [reply]
No, I have not tried yet, it's on my todo list. You could however try to use this as the base image and just add the alltalk installation which did work for me manually in pytorch docker container: https://hub.docker.com/r/pytorch/pytorch

**Reply by vlrevolution** [reply]
The actual Dockerfile can be seen here:
https://github.com/pytorch/pytorch/blob/main/Dockerfile

**Reply by Paladinium** [reply]
I have a docker file for which I intend to create a PR in this project soon for the alltalkbeta branch.

**Reply by RenNagasaki** [reply]
That sounds awesome!

**Reply by Paladinium** [reply]
Here we go: https://github.com/erew123/alltalk_tts/pull/457

The PR yet has to be approved by @erew123 .

Looking forward getting feedback from you.

**Reply by vlrevolution** [reply]
Nice! I will test it tomorrow, thanks for the efforts! 🌞

**Comment by erew123** [comment]
# Docker BETA build

Docker BETA build has been added, thanks to @Paladinium 

The WIKI page on it is here https://github.com/erew123/alltalk_tts/wiki/Docker-Support-(Beta)

This is a work in progress and this page here is the discussion area.

Thanks

**Comment by Demannu** [comment]
Cross referencing for visibility (and my own accountability)
https://github.com/erew123/alltalk_tts/issues/510#issuecomment-2649311065

**Comment by RenNagasaki** [comment]
@vlrevolution I created a vast.ai Template based on the newest docker image from @erew123  and @Paladinium .
[Docker Image](https://hub.docker.com/r/erew123/alltalk_tts)

You can find it here: [Referral Link](https://cloud.vast.ai/?ref_id=152038&creator_id=152038&name=Alltalk_TTS)
**DISCLAIMER: I used an referral Link because it doesn't hurt others, and I get a small benefit. Obviously as soon as @erew123 decides to copy my template or create his own or just asks me to use a non referral Link, I'll change it.**

The important parts are:
1. There's currently a wrong call in the start_finetune.sh script. Simply remove the "-m trainer.distribute --script" from the last line.
2. finetune.py starts with 127.0.0.1 as IP which blocks all access from outside the container. Scroll to the bottom and replace the IP with 0.0.0.0
I created a PR to fix this -> [https://github.com/erew123/alltalk_tts/pull/534](https://github.com/erew123/alltalk_tts/pull/534)

All you have to do is copy your data via cloud sync and move it to your designated area.
Alltalk is installed at /opt/alltalk

**Reply by vlrevolution** [reply]
Wait so to try this on vast.ai, what are the exact steps? Can you explain step by step?

I'll use your affiliate link

**Reply by RenNagasaki** [reply]
Okay @vlrevolution ,
Let's say you want to use it for XTTS.
1. You follow the link, take the template as the one you want to use.
2. (VAST.AI) - Choose a machine (For finetune I recommend a 4090, its the best value/price ratio, + around 50+ GB Ram (more if a big dataset), For inference take whatever floats your boat even a 3060 can easily infer faster than realtime)
3. (VAST.AI) - You wait for it to start.
4. (VAST.AI) - You Cloud Sync your custom data (Dataset + wavs for finetune, or model + voices for inference) into the /workspace folder if you have any
5. (SSH) - You connect via SSH (the big "_connect" button) assuming you already have SSH Keys setup
![grafik](https://github.com/user-attachments/assets/ee4a0819-b706-4497-aad3-b733c4e38be1)

6. (SSH) - You move the data from /workspace to wherever you it (/opt/alltalk/finetune or /opt/alltalk/models voices)
7. (SSH) - (IF Finetune) you edit '/opt/alltalk/start_finetune.sh' with whatever texteditor you like and remove the "-m trainer.distribute --script" from the last line)
8. (SSH) - (If finetune) you edit '/opt/alltalk/finetune.py' with ............................................ and edit the '127.0.0.1' at the bottom of the file to a '0.0.0.0'
9. (SSH) - you call either ./start_finetune.sh or ./start_alltalk.sh to start the corresponding service.
10. (VAST.AI) - You click on the blue ip address at the top of the instance entry 
![grafik](https://github.com/user-attachments/assets/4373622e-ac03-4bbd-be31-06665f942ae7)
There at the bottom you have the ip + mapped ports pointing to your services.
Finetune -> 7052/tcp
Alltalk Gradio -> 7852/tcp
Alltalk API -> 7851/tcp
11. (ALLTALK) - Have Fun!

**Reply by RenNagasaki** [reply]
If you want to use it for piper for example all the steps are the except at the beginning you have to edit the template to use the fitting docker tag (Default is latest-xtts).

**Reply by vlrevolution** [reply]
Thank you brother, that clears everything up. Will try to find some time this weekend and play around with it!

**Reply by vlrevolution** [reply]
> Okay @vlrevolution , Let's say you want to use it for XTTS.
> 
>     1. You follow the link, take the template as the one you want to use.
> 
>     2. (VAST.AI) - Choose a machine (For finetune I recommend a 4090, its the best value/price ratio, + around 50+ GB Ram (more if a big dataset), For inference take whatever floats your boat even a 3060 can easily infer faster than realtime)
> 
>     3. (VAST.AI) - You wait for it to start.
> 
>     4. (VAST.AI) - You Cloud Sync your custom data (Dataset + wavs for finetune, or model + voices for inference) into the /workspace folder if you have any
> 
>     5. (SSH) - You connect via SSH (the big "_connect" button) assuming you already have SSH Keys setup
>        ![grafik](https://private-user-images.githubusercontent.com/10595885/416839931-ee4a0819-b706-4497-aad3-b733c4e38be1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwMTk4NzcsIm5iZiI6MTc0NDAxOTU3NywicGF0aCI6Ii8xMDU5NTg4NS80MTY4Mzk5MzEtZWU0YTA4MTktYjcwNi00NDk3LWFhZDMtYjczM2M0ZTM4YmUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDA5NTI1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxYWJlYmNjMjk0NWQ5MDJkMTQ1OGJlNTAzOTA2M2VhYmM0ZTFjNTMzN2VmMjQ5ZDc1NDlmYzk1MDllZGM1OWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hzdOomZqOY7Lyxusr1LJa8I2sgfXeZYV_GrYbUTSMOU)
> 
>     6. (SSH) - You move the data from /workspace to wherever you it (/opt/alltalk/finetune or /opt/alltalk/models voices)
> 
>     7. (SSH) - (IF Finetune) you edit '/opt/alltalk/start_finetune.sh' with whatever texteditor you like and remove the "-m trainer.distribute --script" from the last line)
> 
>     8. (SSH) - (If finetune) you edit '/opt/alltalk/finetune.py' with ............................................ and edit the '127.0.0.1' at the bottom of the file to a '0.0.0.0'
> 
>     9. (SSH) - you call either ./start_finetune.sh or ./start_alltalk.sh to start the corresponding service.
> 
>     10. (VAST.AI) - You click on the blue ip address at the top of the instance entry
>         ![grafik](https://private-user-images.githubusercontent.com/10595885/416839085-4373622e-ac03-4bbd-be31-06665f942ae7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwMTk4NzcsIm5iZiI6MTc0NDAxOTU3NywicGF0aCI6Ii8xMDU5NTg4NS80MTY4MzkwODUtNDM3MzYyMmUtYWMwMy00YmJkLWJlMzEtMDY2NjVmOTQyYWU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDA5NTI1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiNTU5M2RmZDNhYmQzZTljOTc3ZTg3OWE1M2QyMmIzNjA1NjQ1NDkxYjNiM2VjZDEzODU5YWI1MGE4MzdmM2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZCs4_isk8ojk2HEnK_SXFtd4oytUgTCMqQVJBZTvKKk)
>         There at the bottom you have the ip + mapped ports pointing to your services.
>         Finetune -> 7052/tcp
>         Alltalk Gradio -> 7852/tcp
>         Alltalk API -> 7851/tcp
> 
>     11. (ALLTALK) - Have Fun!

I am sshed into vast.ai machine made from your referral link template but there's no /opt/alltalk and it says command not found: docker if I try to do docker ps. I am misunderstanding something here probably?

**Comment by Paladinium** [comment]
🥳 Great news for those waiting for a Docker image of alltalk v2 beta: Images are now being pushed to [Docker Hub](https://hub.docker.com/r/erew123/alltalk_tts).

To summarize:
- The image should start within seconds not requiring any compilation or downloads
  - This is why there are [Docker tags](https://hub.docker.com/r/erew123/alltalk_tts/tags) for each supported model (currently `xtts`, `piper` and `vits`)
- GPU support is enabled
- DeepSpeed is enabled

As quickstart, just use [docker-start.sh](https://github.com/erew123/alltalk_tts/blob/alltalkbeta/docker-start.sh). There are optional params to pass to the script, e.g. for mounting WAV and RVC files, but also to change config values. All details are described in the [readme](https://github.com/erew123/alltalk_tts/blob/alltalkbeta/DOCKER_README.md) (the [wiki](https://github.com/erew123/alltalk_tts/wiki/Docker-Support-(Beta)) is currently out-of-date).

Of course, if you're an experienced Docker user, you can also directly use `docker run`, passing the arguments you need.

If you are a Docker expert and what to contribute, feel free to open PRs. There are still important tasks and improvements to be done such as:

- ✅ Making the image or `docker-start.sh` more flexible with additional arguments/params/env vars -> https://github.com/erew123/alltalk_tts/pull/557
- Prohibit Gradio to download a file for the theme at startup of the container
- Allowing additional images for other models like f5-tts which would require to extend [firstrun.py](https://github.com/erew123/alltalk_tts/blob/alltalkbeta/system/config/firstrun.py)
- ✅ Cronjob for removing generated WAV files from this when older than x min -> https://github.com/erew123/alltalk_tts/pull/561
- Adding HTTP health check endpoints for [liveness, readiness and startup propes](https://blog.kubecost.com/blog/kubernetes-health-check/)
- Supporting other architectures (e.g. ARM)
- ✅ Enabling [multi-engine-manager](https://github.com/erew123/alltalk_tts/wiki/Multi-Engine-Manager) -> https://github.com/erew123/alltalk_tts/pull/545
- Security: 
  - ✅ Not using `root` to run the application for security reasons -> https://github.com/erew123/alltalk_tts/pull/558
  - Adding vulnerability scanners to the CI
- Supporting other GPUs (e.g. ROCm) (careful with DeepSpeed here - see next line)
- Building and pushing [DeepSpeed](https://github.com/erew123/alltalk_tts/blob/alltalkbeta/docker/deepspeed/Dockerfile) on the CI instead of pulling a [pre-built DeepSpeed wheel](https://github.com/erew123/alltalk_tts/blob/alltalkbeta/Dockerfile#L49) (challenge here is that a GPU and matching CUDA version is required)
- Improving container startup time
- ✅ Improving image build time https://github.com/erew123/alltalk_tts/pull/552
- Improving image size
- Improving documentation
- ...

I won't do this on my own, so any help is appreciated!

**Comment by jbienz** [comment]
When attempting to run the docker image in Ubuntu under WSL (Windows) it appears only port 7851 is mapped to the host and not port 7852. Turns out this is fixed inhttps://github.com/erew123/alltalk_tts/pull/623 but never merged. Can we get this updated?

**Reply by Paladinium** [reply]
@jbienz : Done.

**Reply by jbienz** [reply]
Thank you so much!
